pkg/querier/frontend/worker.go

```diff
@@ -118,116 +111,64 @@ func (w *worker) watchDNSLoop(servCtx context.Context) error {
 			switch update.Op {
 			case naming.Add:
 				level.Debug(w.log).Log("msg", "adding connection", "addr", update.Addr)
-				ctx, cancel := context.WithCancel(servCtx)
-				cancels[update.Addr] = cancel
-				w.runMany(ctx, update.Addr)
+				client, err := w.connect(update.Addr)
+				if err != nil {
+					level.Error(w.log).Log("msg", "error connecting", "addr", update.Addr, "err", err)
+				}
+
+				w.managers[update.Addr] = newFrontendManager(servCtx, w.log, w.server, client, 0, w.cfg.GRPCClientConfig.MaxRecvMsgSize)
 
 			case naming.Delete:
 				level.Debug(w.log).Log("msg", "removing connection", "addr", update.Addr)
-				if cancel, ok := cancels[update.Addr]; ok {
-					cancel()
+				if mgr, ok := w.managers[update.Addr]; ok {
+					mgr.stop()
 				}
 
 			default:
 				return fmt.Errorf("unknown op: %v", update.Op)
 			}
 		}
+
+		w.resetParallelism()
 	}
 }
 
-// runMany starts N runOne loops for a given address.
-func (w *worker) runMany(ctx context.Context, address string) {
-	client, err := w.connect(address)
+func (w *worker) connect(address string) (FrontendClient, error) {
+	opts := []grpc.DialOption{grpc.WithInsecure()}
+	opts = append(opts, w.cfg.GRPCClientConfig.DialOption([]grpc.UnaryClientInterceptor{middleware.ClientUserHeaderInterceptor}, nil)...)
+	conn, err := grpc.Dial(address, opts...)
 	if err != nil {
-		level.Error(w.log).Log("msg", "error connecting", "addr", address, "err", err)
-		return
+		return nil, err
 	}
+	return NewFrontendClient(conn), nil
+}
 
-	w.wg.Add(w.cfg.Parallelism)
-	for i := 0; i < w.cfg.Parallelism; i++ {
-		go w.runOne(ctx, client)
+func (w *worker) resetParallelism() {
```

Comment:
If I read this right:
- When #frontend doesn't divide cleanly into the parallelism limit, we give the remainder as +1s to the last few.
- When #frontend > parallelism limit, we give each worker 1 - potentially exceeding the parallism limit and queueing in the queriers.

I don't see any other alternative at this stage, but I'd like to see comments and warning logs to this effect please.